src/java/org/apache/cassandra/index/sai/disk/v1/InvertedIndexSearcher.java

```diff
@@ -129,11 +153,116 @@ else if (exp.getOp() == Expression.Op.RANGE)
     }
 
+    private Cell<?> readColumn(SSTableReader sstable, PrimaryKey primaryKey)
+    {
+        var dk = primaryKey.partitionKey();
+        var slices = Slices.with(indexContext.comparator(), Slice.make(primaryKey.clustering()));
+        try (var rowIterator = sstable.iterator(dk, slices, columnFilter, false, SSTableReadsListener.NOOP_LISTENER))
+        {
+            var unfiltered = rowIterator.next();
+            assert unfiltered.isRow() : unfiltered;
+            Row row = (Row) unfiltered;
+            return row.getCell(indexContext.getDefinition());
+        }
+    }
+
     @Override
     public CloseableIterator<PrimaryKeyWithSortKey> orderBy(Orderer orderer, Expression slice, AbstractBounds<PartitionPosition> keyRange, QueryContext queryContext, int limit) throws IOException
     {
-        var iter = new RowIdWithTermsIterator(reader.allTerms(orderer.isAscending()));
-        return toMetaSortedIterator(iter, queryContext);
+        if (!orderer.isBM25())
+        {
+            var iter = new RowIdWithTermsIterator(reader.allTerms(orderer.isAscending()));
+            return toMetaSortedIterator(iter, queryContext);
+        }
+        if (docLengthsReader == null)
+            throw new InvalidRequestException(indexContext.getIndexName() + " does not support BM25 scoring until it is rebuilt");
+
+        // find documents that match each term
+        var queryTerms = orderer.getQueryTerms();
+        var postingLists = queryTerms.stream()
+                                     .collect(Collectors.toMap(Function.identity(), term ->
+        {
+            var encodedTerm = version.onDiskFormat().encodeForTrie(term, indexContext.getValidator());
+            var listener = MulticastQueryEventListeners.of(queryContext, perColumnEventListener);
+            var postings = reader.exactMatch(encodedTerm, listener, queryContext);
+            return postings == null ? PostingList.EMPTY : postings;
+        }));
+        // extract the match count for each
+        var documentFrequencies = postingLists.entrySet().stream().collect(Collectors.toMap(Map.Entry::getKey, e -> (long) e.getValue().size()));
+
+        var pkm = primaryKeyMapFactory.newPerSSTablePrimaryKeyMap();
+        var merged = IntersectingPostingList.intersect(postingLists);
+        
+        // Wrap the iterator with resource management
+        var it = new AbstractIterator<DocTF>() { // Anonymous class extends AbstractIterator
+            private boolean closed;
+            
+            @Override
+            protected DocTF computeNext()
+            {
+                try
+                {
+                    int rowId = merged.nextPosting();
+                    if (rowId == PostingList.END_OF_STREAM)
+                        return endOfData();
+                    int docLength = docLengthsReader.get(rowId);
+                    return new DocTF(pkm.primaryKeyFromRowId(rowId), docLength, merged.frequencies());
```

Comment:
We get segment row ids from the posting list and need to pass sstable row ids to the pkm.